### PR TITLE
Re-implementation of Tang (1990) log procedures in pure float32

### PR DESCRIPTION
**Pull request by sotlampr**
`Base.Math.log1p` calculation relies on two procedures , [log_proc1](https://github.com/JuliaLang/julia/blob/15b590bd6a1f17f3859752594d5d8ff534cd38b3/base/special/log.jl#L222) and [log_proc2](https://github.com/JuliaLang/julia/blob/15b590bd6a1f17f3859752594d5d8ff534cd38b3/base/special/log.jl#L245) from the literature [Tang (1990)](https://dl.acm.org/doi/pdf/10.1145/98267.98294). The current implementation upscales to `Float64` internally and as such is incompatible with Metal.

This PR attempts to convert the two procedures to pure-float32.

See also relevant discussion #234

### CODE REVIEW

File: src/device/intrinsics/math.jl [hunk 1]
```diff
@@ -101,6 +102,80 @@ using Base: FastMath
...
+
+@inline logb(::Val{2})  = 1.442695f0
+@inline logb(::Val{:ℯ}) = 1f0
+@inline logb(::Val{10}) = 0.4342945f0
+
+@device_override function Base.Math.log_proc1(y::Float32,mf::Float32,F::Float32,f::Float32,base=Val(:ℯ))
+    jp = unsafe_trunc(Int,128.0f0*F)-127
+
+    ## Steps 1 and 2
+    hi = t_log_Float32[jp]
```
Comment (sotlampr) [inline src/device/intrinsics/math.jl:136]: Julia Base uses `Float64` tables

File: src/device/intrinsics/math.jl [hunk 2]
```diff
@@ -101,6 +102,80 @@ using Base: FastMath
...
+    #                 Float32(-0x1.00006p-1),
+    #                 Float32(0x1.55546cp-2))
+
+    ## Step 3' (alternative)
+    u = (2f0f)/(y+F)
+    v = u*u
+    q = u*v*0.08333351f0
+
+    ## Step 4
+    logb(base)*(l + (u + q))
```
Comment (sotlampr) [inline src/device/intrinsics/math.jl:151]: Similarly, `logb` in base returns a double

File: src/device/intrinsics/math.jl [hunk 3]
```diff
@@ -101,6 +102,80 @@ using Base: FastMath
...
+    u = 2(f*g)
+    v = u*u
+
+    ## Step 2
+    q = u*v*@horner(v,
+                    0.08333332f0,
+                    0.012512346f0)
+
+    ## Step 3
+    @inline function truncate(x)
```
Comment (sotlampr) [inline src/device/intrinsics/math.jl:166]: From Tang (1990):

> (1) Define M := 12 for single precision, and define M := 24 for double precision.
(2) u1 := u rounded (or truncated) to M significant bits.
(3) fi := f rounded (or truncated) to M significant bits.

I'm not sure if this is what is meant

File: src/device/intrinsics/math.jl [hunk 4]
```diff
@@ -101,6 +102,80 @@ using Base: FastMath
...
+      reinterpret(Float32,
+                  reinterpret(Int32, 0.012512346f0) & 0b11111111111100000000000000000000)
+    end
+    u₁ = truncate(u)
```
Comment (sotlampr) [inline src/device/intrinsics/math.jl:170]: Non-upcasting alternative procedure from Tang (1990)

**Comment by maleadt** [comment]
cc @oscardssmith

**Comment by oscardssmith** [comment]
What's the error like for these implementations?

**Comment by sotlampr** [comment]
This `log1p` compared with `Base.Math.log1p` (`Float32`) (ULP):
```
max 4.0 at x = -0.099609345
mean 0.12487844518189703
```

This `log_proc2` compared with `Base.Math.log_proc2` (`Float32`) ( ULP):
```
log_proc2{Float32}, base=2
max 2.0 at x = 0.043909933
mean 0.1784229523275927

log_proc2{Float32}, base=ℯ
max 1.0 at x = 0.064494334
mean 0.004462545515167737

log_proc2{Float32}, base=10
max 2.0 at x = 0.03663287
mean 0.287801184326243
```

**Comment by sotlampr** [comment]
**NOTE** This doesn't work on metal - the result are zeros.

**Comment by oscardssmith** [comment]
Can you compare against Base.Math.log1p (Float64)? your current comparison isn't great since Base.Math.log1p (Float32) also has rounding error.

**Comment by sotlampr** [comment]
I assumed that we're interested in the difference between the upcasting float32 vs. the pure float32 implementation.

This `log1p{Float32}` v. `Base.Math.log1p{Float64}`
```
max 4.441346645355225 at x = -0.11138753
mean 0.17755624519091973
```

And for reference, `Base.Math.log1p{Float32}` v. `Base.Math.log1p{Float64}`
```
max 0.563554048538208 at x = 0.08200329
mean 0.11855523097385935
```

**Comment by sotlampr** [comment]
@oscardssmith do you think the error is acceptable? And most critically, running on Metal does not work - the error is way too high. Can you see any possible culprit in the code?

**Comment by oscardssmith** [comment]
the error is probably acceptable, but it is higher than I was expecting. I don't know why it's giving wrong answers on gpu. possibly the truncate is misbehaving?

**Comment by oscardssmith** [comment]
now that you mention it, a table based method is probably not the best idea for a gpu.

**Comment by oscardssmith** [comment]
https://github.com/JuliaMath/openlibm/blob/master/src/s_log1pf.c might be a better approach.

**Comment by sotlampr** [comment]
Replaced by #239

**Pull request closed without merging**